### PR TITLE
mgr/cephadm: pick correct IPs for ingress service based on VIP

### PR DESCRIPTION
**Pull request by adk3798**
We need to make sure the keepalive conf sets
the unicast src and peer IPs to be the ones
in the same subnet as the VIP we're setting up,
as well as specify the correct interface. Otherwise,
the keepalive daemons don't speak to each other
properly and all end up going into MASTER state.
    
Fixes: https://tracker.ceph.com/issues/62276
    
Signed-off-by: Adam King <adking@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/services/ingress.py [hunk 1]
```diff
@@ -341,22 +320,51 @@ def keepalived_generate_config(
...
+        vrrp_interfaces: List[str] = []
+        if not spec.vrrp_interface_network:
+            vrrp_interfaces = interfaces
+        else:
+            for subnet, ifaces in self.mgr.cache.networks.get(host, {}).items():
+                if subnet == spec.vrrp_interface_network:
+                    vrrp_interface = [list(ifaces.keys())[0]] * len(interfaces)
+                    logger.info(
+                        f'vrrp will be configured on {host} interface '
+                        f'{vrrp_interface} (which has guiding subnet {subnet})'
```
Comment (phlogistonjohn) [inline src/pybind/mgr/cephadm/services/ingress.py:346]: guiding?

File: src/pybind/mgr/cephadm/serve.py [hunk 2]
```diff
@@ -695,8 +696,8 @@ def _apply_service(self, spec: ServiceSpec) -> bool:
                 public_networks = [x.strip() for x in out.split(',')]
                 self.log.debug('mon public_network(s) is %s' % public_networks)
 
-        def matches_network(host):
-            # type: (str) -> bool
+        def matches_public_network(host, sspec):
+            # type: (str, ServiceSpec) -> bool
```
Comment (phlogistonjohn) [inline src/pybind/mgr/cephadm/serve.py:700]: As you advised me the other day, if you're touching the function signature, convert it to the full type annotations form.

File: src/pybind/mgr/cephadm/serve.py [hunk 3]
```diff
@@ -713,6 +714,40 @@ def matches_network(host):
...
+            virtual_ips = []
+            if ingress_spec.virtual_ip:
+                virtual_ips.append(ingress_spec.virtual_ip)
+            elif ingress_spec.virtual_ips_list:
+                virtual_ips = ingress_spec.virtual_ips_list
+            for vip in virtual_ips:
+                found = False
+                bare_ip = str(vip).split('/')[0]
+                for subnet, ifaces in self.mgr.cache.networks.get(host, {}).items():
+                    if ifaces and ipaddress.ip_address(bare_ip) in ipaddress.ip_network(subnet):
```
Comment (phlogistonjohn) [inline src/pybind/mgr/cephadm/serve.py:732]: I feel like I keep seeing this same loop. Not for this PR, per se, but maybe we should try to encapulate this common logic in a function, possibly one that takes a predicate method for filtering.

**Review by phlogistonjohn — COMMENTED** [review]
Generally this looks OK to me but I'll admit that I find it a bit difficult to follow. In the future consider writing more "bite size" functions that are easier to understand (and test!) and then compose the higher level logic out of those.

File: src/pybind/mgr/cephadm/services/ingress.py [hunk 4]
```diff
@@ -247,56 +247,35 @@ def keepalived_generate_config(
...
-        bare_ips = []
-        if spec.virtual_ip:
-            bare_ips.append(str(spec.virtual_ip).split('/')[0])
-        elif spec.virtual_ips_list:
-            bare_ips = [str(vip).split('/')[0] for vip in spec.virtual_ips_list]
-        interface = None
-        for bare_ip in bare_ips:
+        def _get_valid_interface_and_ip(vip: str, host: str) -> Tuple[str, str]:
+            # interface
+            bare_ip = str(vip).split('/')[0]
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply src/pybind/mgr/cephadm/services/ingress.py:252]: Correcting myself now that I see the tests and understand the inputs better: we'd want `ipaddress.ip_interface(...)` for this use case not `ipaddress.ip_network`.

```python
>>> ipaddress.ip_interface('192.168.44.66/24').ip
IPv4Address('192.168.44.66')
>>> ipaddress.ip_interface('192.168.44.66').ip
IPv4Address('192.168.44.66')
```

My general point stands, let's rely on the library for as much of the heavy lifting as possible.

File: src/pybind/mgr/cephadm/services/ingress.py (hunk 1 above)
Comment (adk3798) [reply src/pybind/mgr/cephadm/services/ingress.py:346]: this was just copy-pasted so I'm not sure exactly what "guiding" means here. I'll just remove it as part of this work.

File: src/pybind/mgr/cephadm/serve.py (hunk 3 above)
Comment (adk3798) [reply src/pybind/mgr/cephadm/serve.py:732]: made a tracker for this so we can cover it in a follow up https://tracker.ceph.com/issues/62327

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2023-08-15_13:44:09-orch:cephadm-wip-adk-testing-2023-08-14-1902-distro-default-smithi/

7 failures, 3 dead (as of writing this 1 dead and 2 running but the 2 running will time out) jobs

- 4 failures were mds_upgrade_sequence upgrading from reef -> main. The upgrade from reef doesn't work yet as we need to implement something that can pull the latest reef build of the cephadm binary
- 2 failures were https://tracker.ceph.com/issues/62084, known issue
- 1 failure was https://tracker.ceph.com/issues/59704, known issue
- 2 dead jobs were mds upgrade sequence starting from quincy. They time out doing a `umount` command. These were broken the same way before when upgrading from pacific, and we wanted to see if upgrading from quincy would fix it somehow but it hasn't.
- 1 dead job was infra/teuthology issue: `Error reimaging machines: Expected smithi172's OS to be centos 8 but found rhel 8.6`

Overall, nothing related to the mds upgrade sequence or  jaeger-tracing can be merged, but nothing here to block merging most other things in the run.

**Review by phlogistonjohn — APPROVED** [review]
Looks OK to me.

**Pull request merged**